pkg/controller/meta_pod_controller.go

```diff
@@ -0,0 +1,231 @@
+
+// MetaPodController reconciles meta pods object.
+type MetaPodController struct {
+	client.Client
+}
+
+// +kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch;update;patch
+
+// getMetaRole sends a MetaMember request to a meta node at ip:port, determining if the node is a leader. Returns true if operation was successful.
+func (mpc *MetaPodController) getMetaRole(ctx context.Context, host string, port uint, podName string) (string, bool) {
+	log := log.FromContext(ctx)
+	addr := fmt.Sprintf("%s:%v", host, port)
+
+	newCtx, cancel := context.WithTimeout(ctx, time.Second)
+	defer cancel()
+	conn, err := grpc.DialContext(newCtx, addr, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		log.Error(err, fmt.Sprintf("Unable to connect to meta pod at %s", addr))
+		return "", false
+	}
+	defer conn.Close()
+	c := pb.NewMetaMemberServiceClient(conn)
+
+	resp, err := c.Members(ctx, &pb.MembersRequest{})
+	if err != nil {
+		log.Info(fmt.Sprintf("Sending MembersRequest failed. Assuming meta node is not yet ready. Error was: %s", err.Error()))
+		return consts.MetaRoleUnknown, true
+	}
+
+	if len(resp.Members) == 0 {
+		return consts.MetaRoleUnknown, true
+	}
+
+	// Assuming member.Address.Host like risingwave-meta-default-0.risingwave-meta
+	for _, member := range resp.Members {
+		memberAddress := ""
+		if arr := strings.Split(member.Address.Host, "."); len(arr) == 2 {
+			memberAddress = arr[0]
+		} else {
+			log.Error(fmt.Errorf("unexpected member format %s", member.Address.Host), "")
+			return consts.MetaRoleUnknown, false
+		}
+		if member.IsLeader && podName == memberAddress && port == uint(member.Address.Port) {
+			return consts.MetaRoleLeader, true
+		}
+	}
+	return consts.MetaRoleFollower, true
+}
+
+// getMetaPort returns the service port of the pod. Returns true if operation was successful.
+func getMetaPort(pod *corev1.Pod) (uint, bool) {
+	for _, container := range pod.Spec.Containers {
+		if container.Name == "meta" {
+			for _, containerPort := range container.Ports {
+				if containerPort.Name == consts.PortService {
+					return uint(containerPort.ContainerPort), true
+				}
+			}
+		}
+	}
+	return 0, false
+}
+
+// getPodIP returns the pod IP of the pod. Returns true if operation was successful.
+func getPodIP(pod *corev1.Pod) (string, bool) {
+	ip := pod.Status.PodIP
+	return ip, ip != ""
+}
+
+// getOtherMetPods returns all other meta pods of the same RW instance.
+func (mpc *MetaPodController) getOtherMetPods(metaPod *corev1.Pod, ctx context.Context) ([]corev1.Pod, error) {
+	log := log.FromContext(ctx)
+
+	// get other meta pods of this RW instance
+	otherMetaPods := &corev1.PodList{}
+	labelSet := map[string]string{consts.LabelRisingWaveComponent: consts.ComponentMeta}
+	if rwInstance, ok := metaPod.ObjectMeta.Labels[consts.LabelRisingWaveName]; ok {
+		labelSet[consts.LabelRisingWaveName] = rwInstance
+	} else {
+		log.Error(fmt.Errorf("unable to retrieve risingwave name from pod"), "")
```

Comment:
It should break here and return an empty list. Otherwise, the function scans every Pod with the `risingwave/role=meta` label.